src/security/owasp-zap/tasks/create-bugs.yaml

```diff
@@ -0,0 +1,118 @@
+# Create bugs for any failed tests
+parameters:
+  - name: createBugsForFailures # OPTIONAL: If the OWASP pipeline fails, have Azure DevOps create bug cards
+    default: false
+  - name: parentCardId # OPTIONAL: If 'createBugsForFailures' is true, provide a parent card number for any raised cards
+    default: ''
+  - name: organisationUri # OPTIONAL: The URI of the DevOps organisation to create the bugs in. Defaults to the current organisation.
+    default: '$(System.TeamFoundationCollectionUri)'
+  - name: projectName # OPTIONAL: The name of the DevOps project to create the bugs in. Defaults to the current project.
+    default: '$(System.TeamProject)'
+  - name: name # REQUIRED: The name of the scan
+    type: string
+
+steps:
+  - task: PowerShell@2
+    displayName: 'Create Bug Cards for OWASP ZAP Failures'
+    condition: and(succeeded(), eq('${{ parameters.createBugsForFailures }}', 'true'))
+    env:
+      AZURE_DEVOPS_EXT_PAT: $(System.AccessToken) # Used by az devops commands
+    inputs:
+      targetType: inline
+      script: |
+        filter ConvertTo-FoldedString { $_ -replace "\r?\n", " " }
+
+        $organization = "${{parameters.organisationUri}}"
+        $projectName = "${{parameters.projectName}}"
+        $parentCardId = "${{parameters.parentCardId}}"
+        $scanName = "${{parameters.name}}"
+
+        # Load the ZAP alerts from the Report XML
+        $reportPath = "$(workingTempDirectory)/Report.xml"
+        [xml]$zapReport = Get-Content -Path $reportPath
+
+        $failedScenarios = $zapReport.OWASPZAPReport.site.alerts.alertitem
+
+        # Loop through each failed scenario
+        foreach ($scenario in $failedScenarios) {
+          # Define a bug title in the format "Failing Scenario Name (Risk Level)"
+          # Note: Sometimes the risk descriptions are duplicative e.g. "Medium (Medium)"
+          $bugTitle = "OWASP Issue: $($scenario.name) $($scenario.riskdesc)"
+
+          Write-Host ""
+          Write-Host "Processing bug '$bugTitle'"
+
+          # Query DevOps to check if this bug already exists in an active state
+          $workItemQuery = @"
+          SELECT [System.Id], [System.Title]
+          FROM WorkItems
+          WHERE
+            [System.TeamProject] = '$projectName' AND
+            [System.WorkItemType] = 'Bug' AND
+            [System.State] <> 'Removed' AND
```

Comment:
Are there other statuses that should be excluded, e.g. 'Done'? Just thinking about the situation whereby a big gets raised and resolved, but then is reintroduced in a subsequent change.

Reply:
I think "Done" and "Removed" will be fine as defaults. How hard would it be to parameterise this (defaulting to those two statuses)? As some projects may use different statuses. No worries if it's loads of rework.

Reply:
Agreed to not parameterise.